ruby_event_store/lib/ruby_event_store/mappers/transformation/preserve_types.rb

```diff
@@ -0,0 +1,132 @@
+# frozen_string_literal: true
+
+module RubyEventStore
+  module Mappers
+    module Transformation
+      SerializationError = Class.new(ArgumentError)
+
+      class PreserveTypes
+        def initialize
+          @registered_type_serializers = {}
+        end
+
+        def register(type, serializer:, deserializer:)
+          @registered_type_serializers[type.to_s] = {
+            serializer: serializer,
+            deserializer: deserializer
+          }
+          self
+        end
+
+        def dump(record)
+          types = store_types(record.data)
+
+          Record.new(
+            event_id:   record.event_id,
+            event_type: record.event_type,
+            data:       transform_data(record.data),
+            metadata:   record.metadata.merge(types: types),
+            timestamp:  record.timestamp,
+            valid_at:   record.valid_at,
+          )
+        end
+
+        def load(record)
+          types = record.metadata.delete(:types)
+          data = types ? restore_types(record.data, types) : record.data
+
+          Record.new(
+            event_id:   record.event_id,
+            event_type: record.event_type,
+            data:       data,
+            metadata:   record.metadata,
```

Comment:
Why metadata isn't transformed? Timestamps and simple value objects (`RequestUuid`, `UserAgent` or just DateTime objects) may exist there too

Reply:
Right... adding that.